.github/workflows/validate-migration.yaml

```diff
@@ -73,29 +171,43 @@ jobs:
           mkdir -p vendor/helix-core-api/linux
           tar -C vendor/helix-core-api/linux -xzf p4.tgz --strip 1
 
-      - name: "Cache OpenSSL"
+      - name: "Cache OpenSSL source"
+        id: cache-openssl-source
         uses: actions/cache@v2
-        id: cache-openssl
         with:
-          path: ${{ env.OPENSSL_INSTALL_DIR }}
-          key: ${{ runner.os }}-openssl-${{ env.OPENSSL_DOWNLOAD_URL }}
+          path: ${{ env.OPENSSL_SOURCE_DIR }}
+          key: ${{ runner.os }}-openssl-source-${{ env.OPENSSL_DOWNLOAD_URL }}
 
-      - name: "Compile OpenSSL"
-        if: ${{ steps.cache-openssl.outputs.cache-hit != 'true' }}
+      - name: "Download OpenSSL source"
+        if : ${{ steps.cache-openssl-source.outputs.cache-hit != 'true' }}
         run: |
-          mkdir openssl-src
           curl -L "${OPENSSL_DOWNLOAD_URL}" --output openssl.tgz
-          tar -C openssl-src -xzf openssl.tgz --strip 1
-          pushd openssl-src
-          ./config --prefix="${OPENSSL_INSTALL_DIR}" --static
-          make install
-          popd 
+          mkdir -p "${OPENSSL_SOURCE_DIR}" || true
+          tar -C "${OPENSSL_SOURCE_DIR}" -xzf openssl.tgz --strip-components 1
+          rm openssl.tgz || true
+
+      - name: "Compile OpenSSL"
+        run: |
+          .github/workflows/validate-migration-scripts/compile-openssl.sh
+        env:
+          CC: "sccache clang"
+          CXX: "sccache clang++"
 
       - name: "Run validate migration script"
         run: |
-          sudo apt-get install gettext --yes # for envsubst 
-          .github/workflows/run-test.sh
+          .github/workflows/validate-migration-scripts/run-test.sh
         env:
-          P4CLIENT: "validate-migration-${{ github.ref_name }}-${{ github.run_number }}-${{ matrix.sanitizer }}" # the temporary Perforce client name to use while validating the git migration
+          P4CLIENT: "validate-migration-${{ github.ref_name }}-${{ github.sha }}-${{ github.run_number }}-${{ matrix.tool }}" # the temporary Perforce client name to use while validating the git migration
           OPENSSL_ROOT_DIR: ${{ env.OPENSSL_INSTALL_DIR }}
-          CMAKE_CXX_FLAGS: "${{ matrix.CMAKE_CXX_FLAGS }} -O0"
\ No newline at end of file
+          CMAKE_CXX_COMPILER: "clang++"
+          CMAKE_C_COMPILER: "clang"
+          CMAKE_C_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_COMPILER_LAUNCHER: "sccache"
+          CMAKE_CXX_FLAGS: "${{ env.CMAKE_CXX_FLAGS }} -stdlib=libc++ -fno-omit-frame-pointer -O0"
+          UBSAN_OPTIONS: "print_stacktrace=1" # print stack trace on undefined behavior
+          ASAN_OPTIONS: "check_initialization_order=1" # check initialization order of globals
+
+      - name: "Ensure p4 client is deleted"
+        if: always()
+        run: |
+          p4 client -f -Fs -d "${P4CLIENT}" || true
```

Comment:
Is the client started with a timeout or something? Is this step guaranteed to run even if someone cancels a running job midway using the web UI (and if so, would that be bad?)?

Reply:
Our perforce license allows for a limited number of [clients](https://www.perforce.com/manuals/p4guide/Content/P4Guide/configuration.workspace.html) to be defined at any given time (~20). For each job in the matrix, we creature a unique client (with a name that's unique to each job using a mix of the job number, tool we're using, sha, etc.). 

I have [logic in the run-test script](https://github.com/sourcegraph/p4-fusion/blob/94cc616797d2bf713ffbbb27e4625ecd253f87ee/.github/workflows/validate-migration-scripts/run-test.sh#L21-L42) to ensure that we delete the client whenever the script exits using a trap. However, if the job as a whole is cancelled, there is no guaranteed that trap will execute (leaving behind a client that consumes one of our licenses). This extra step that I added here will _always_ run (through the use of the `if: always()` even if the job is cancelled - ensuring that the client is cleaned up in most circumstances.

Reply:
>Is this step guaranteed to run even if someone cancels a running job midway using the web UI (and if so, would that be bad?)?

Yes, it is guaranteed to run according to https://docs.github.com/en/actions/learn-github-actions/expressions#always. This is the explicit intent (to ensure that we don't consume one of the clients on our license). This is not bad, since we create a new client for each commit / job.